### PR TITLE
Trip Details: Add react-intl and TypeScript

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
This PR adds react-intl and TypeScript to the TripDetails package. A new `types` packages is introduced to hold OTP-UI -specific types shared between multiple packages.

The PR contains breaking changes in the TripDetails props. Specifically, implementers must wrap the component with `<IntlProvider>` from react-intl.  The `messages` and `timeOptions` props are dropped, replaced with customizable messages that implementers pass to `<IntlProvider>`. Default English and French message files are provided.  On the other hand, ~a new `currency` props is added~=> EDIT: removed in https://github.com/opentripplanner/otp-ui/pull/281/commits/506a8bd21c74ba893af7e8ce5aa1e9389f1c7113, and slots have been added to customize the expandable details section of departure, fare, and calories section.

package.json and other configurations are tweaked to accommodate these changes.

EDIT: This PR is a partial solution to #237.

image

### CODE REVIEW

**Comment by binh-dam-ibigroup** [comment]
There is potential to remove the `currency` prop that is introduced in this PR if #282 is adopted, and use that PR's currency calculation logic instead.

File: .github/workflows/codespell.yml
Comment (evansiroky) [inline .github/workflows/codespell.yml:16]: Is there a way to just ignore only the non-english i18n files?

File: packages/types/package.json
Comment (evansiroky) [inline packages/types/package.json:4]:
```suggestion
  "description": "TypeScript types used across multiple OTP-UI packages",
```

File: packages/trip-details/i18n/en-US.yml
Comment (evansiroky) [inline packages/trip-details/i18n/en-US.yml:2]: Just a suggestion

```suggestion
# To use from a react-intl application, merge the content of this file into the messages object
```

File: packages/trip-details/i18n/en-US.yml
Comment (evansiroky) [inline packages/trip-details/i18n/en-US.yml:4]: I think it'd look a lot nicer to write out the hierarchy of these and then flatten in the code.

File: packages/trip-details/package.json
Comment (evansiroky) [inline packages/trip-details/package.json:14]: This might result in only this folder and not the `lib` or `esm` folders being included in the release. What was the intention of putting this in here? Perhaps there is a different way to accomplish that.

File: packages/trip-details/src/TripDetails.story.tsx
Comment (evansiroky) [inline packages/trip-details/src/TripDetails.story.tsx:3]: The [@storybook/addon-knobs](https://www.npmjs.com/package/@storybook/addon-knobs) package is deprecated. Please use [Storybook args and argTypes](https://storybook.js.org/docs/react/writing-stories/args) instead.

File: packages/trip-details/src/TripDetails.story.tsx
Comment (evansiroky) [inline packages/trip-details/src/TripDetails.story.tsx:95]: The [@storybook/addon-knobs](https://www.npmjs.com/package/@storybook/addon-knobs) package is deprecated. Please use [Storybook args and argTypes](https://storybook.js.org/docs/react/writing-stories/args) instead.

File: packages/trip-details/src/TripDetails.story.tsx
Comment (evansiroky) [inline packages/trip-details/src/TripDetails.story.tsx:119]: The [@storybook/addon-knobs](https://www.npmjs.com/package/@storybook/addon-knobs) package is deprecated. Please use [Storybook args and argTypes](https://storybook.js.org/docs/react/writing-stories/args) instead.

File: packages/trip-details/src/index.tsx
Comment (evansiroky) [inline packages/trip-details/src/index.tsx:22]: Since we're assuming the use of react-intl, I think we shouldn't bother using default messages.

File: packages/trip-details/src/index.tsx
Comment (evansiroky) [inline packages/trip-details/src/index.tsx:12]: Since we're assuming the use of react-intl, I think we shouldn't bother using default messages.

File: packages/trip-details/src/index.tsx
Comment (evansiroky) [inline packages/trip-details/src/index.tsx:22]: This is creating a fair amount of duplication and long-term need to update this object every time a new message is created. Perhaps it'd be better to make a function that assists as follows:

```suggestion
function makeMessageId (suffix: string) {
  return `otpUi.TripDetails.${suffix}`;
}
```

File: packages/trip-details/src/index.tsx
Comment (evansiroky) [inline packages/trip-details/src/index.tsx:46]: I think it'd actually be cleaner to adhere to the linting rule and put the following anonymous function into a named function. This should help with debugging.

File: packages/trip-details/src/index.tsx
Comment (evansiroky) [inline packages/trip-details/src/index.tsx:47]: In order to show that more than just an `a` tag is used here, I think it'd be good to rename this to something like `dietaryLink`.

File: packages/trip-details/src/index.tsx
Comment (evansiroky) [inline packages/trip-details/src/index.tsx:102]: Gosh, this seems like a poor prop choice on the part of react-intl. Is there an alias we can use?

File: packages/trip-details/src/TripDetails.story.tsx
Comment (evansiroky) [inline packages/trip-details/src/TripDetails.story.tsx:95]: Can the `overrideDefaultMessages` be renamed to `useCustomMessages`?

**Review by evansiroky — CHANGES_REQUESTED** [review]
See various comments, but I think we are definitely heading down the right track here finally. I really like how this is shaping up.

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx:12]: I think default messages are still useful if implementers don't provide their own messages (e.g. while getting started).

File: packages/trip-details/src/TripDetails.story.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/TripDetails.story.tsx]: Updated in https://github.com/opentripplanner/otp-ui/pull/281/commits/da0eb14bbfb78531b95e966f84d82529b0c53db5.

File: packages/trip-details/src/TripDetails.story.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/TripDetails.story.tsx]: Updated in https://github.com/opentripplanner/otp-ui/pull/281/commits/da0eb14bbfb78531b95e966f84d82529b0c53db5.

File: packages/trip-details/src/TripDetails.story.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/TripDetails.story.tsx]: Updated in https://github.com/opentripplanner/otp-ui/pull/281/commits/da0eb14bbfb78531b95e966f84d82529b0c53db5.

File: packages/trip-details/src/TripDetails.story.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/TripDetails.story.tsx]: Updated in https://github.com/opentripplanner/otp-ui/pull/281/commits/da0eb14bbfb78531b95e966f84d82529b0c53db5.

File: .github/workflows/codespell.yml
Comment (binh-dam-ibigroup) [reply .github/workflows/codespell.yml:16]: [I tried here](https://github.com/opentripplanner/otp-react-redux/blob/9770827d47c39ccc9103958790a402acc8f32788/.github/workflows/codespell.yml#L15), but the pattern `./i18n/!(en-US.yml)` isn't recognized for some reason (suggestions welcome).

File: packages/trip-details/package.json
Comment (binh-dam-ibigroup) [reply packages/trip-details/package.json]: Yeah that was not necessary. Removed in https://github.com/opentripplanner/otp-ui/pull/281/commits/51d8398e9227f98c910f7b20db5363a707c5f97a.

File: packages/trip-details/i18n/en-US.yml
Comment (binh-dam-ibigroup) [reply packages/trip-details/i18n/en-US.yml]: Updated in https://github.com/opentripplanner/otp-ui/pull/281/commits/21beebcf571549e35ac87db6072146d47d737417, ended up looking better that way!

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: I ended up using a slightly different approach in https://github.com/opentripplanner/otp-ui/pull/281/commits/21beebcf571549e35ac87db6072146d47d737417, but that big block is gone.

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: Updated in https://github.com/opentripplanner/otp-ui/pull/281/commits/8940535c0c10fc9e77c5aaac8b44e350cf04eff9.

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: Updated in https://github.com/opentripplanner/otp-ui/pull/281/commits/8940535c0c10fc9e77c5aaac8b44e350cf04eff9.

File: __snapshots__/storybook.test.ts.snap
Comment (evansiroky) [inline __snapshots__/storybook.test.ts.snap:48390]: These should not be deleted.

File: packages/trip-details/src/index.tsx
Comment (evansiroky) [inline packages/trip-details/src/index.tsx:41]: I'm kind of wondering if all of this is really necessary. It is possible to simply provide string literals for the message IDs and default messages in each `FormattedMessage` call. I suppose this approach might help with static typing to make sure the messages are coded correctly, but there is still no guarantee that the imported data from the messages yml file will have exact matches (especially with the default messages which could prove to be a problem if that fallback were to occur). There does exist some helpful [tooling](https://formatjs.io/docs/tooling/cli) that can automatically detect message definitions and also the default messages which could then be compared against what exists in the yml files. Additionally, removing this might simplify things a bit for downstream implementers so that they aren't required to input yml files in their transpilation workflow.

File: packages/trip-details/src/TripDetails.story.tsx
Comment (evansiroky) [inline packages/trip-details/src/TripDetails.story.tsx:17]: To reduce confusion with default messages for react-intl, I'd simply call these as follows:

```suggestion
import englishMessages from "../i18n/en-US.yml";
import frenchMessages from "../i18n/fr.yml";
```

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx:41]: I felt it was necessary to collect message ids because each id is used twice in each `FormattedMessage`, once to set the id of that message, and once to set the default English message.

File: packages/trip-details/src/TripDetails.story.tsx
Comment (evansiroky) [inline packages/trip-details/src/TripDetails.story.tsx:141]: This might be what was causing all of those Storyshot snapshots to disappear. I think Storyshot might be unable to process dynamically generated stories like this.

File: packages/trip-details/src/index.tsx
Comment (evansiroky) [reply packages/trip-details/src/index.tsx:41]: Yeah, but the point still stands that if the messages aren't actually defined in the yml file, then it won't work properly. Furthermore, now that I think about it, this is probably incompatible with how the formatjs cli extracts message IDs (see [here](https://formatjs.io/docs/tooling/cli)).

File: packages/trip-details/src/TripDetails.story.tsx
Comment (evansiroky) [inline packages/trip-details/src/TripDetails.story.tsx:186]: Much better, looks great!

**Review by evansiroky — CHANGES_REQUESTED** [review]
We shouldn't delete the storyshot snapshots and we should come to an agreement about how to set the default messages and make message IDs.

File: packages/trip-details/src/TripDetails.story.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/TripDetails.story.tsx:141]: I spent some time replaying the commits of this PR, and a change in https://github.com/opentripplanner/otp-ui/pull/281/commits/6ef1c8493b2073386b830a10020dc237dd158ca6 is causing storyshot to throw out all of these snapshots. Digging further, it appears that importing the YML files is what is causing that, so we have to check that loading YML files is supported in storyshot.

Comment (binh-dam-ibigroup) [reply packages/trip-details/src/TripDetails.story.tsx:141]: This one was really tricky, I have added a YML loader, tried to play well with other code, and updated the snapshots in https://github.com/opentripplanner/otp-ui/pull/281/commits/9725b4bf3d43303d383b1d42219fe199c3b7601e.

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: Partially addressed by https://github.com/opentripplanner/otp-ui/pull/281/commits/9725b4bf3d43303d383b1d42219fe199c3b7601e, bu I suspect we should use the `defineMessages` API?

File: __snapshots__/storybook.test.ts.snap
Comment (binh-dam-ibigroup) [reply __snapshots__/storybook.test.ts.snap]: Snapshots reinstated and updated in https://github.com/opentripplanner/otp-ui/pull/281/commits/9725b4bf3d43303d383b1d42219fe199c3b7601e.

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: `defineMessages` API added in https://github.com/opentripplanner/otp-ui/pull/281/commits/298ffc9e50465de75a7f4c365fc3f12e6db9af16.

**Comment by binh-dam-ibigroup** [comment]
Okay, probably ready for review again with TypeScript fully in force now.

File: packages/trip-details/src/index.tsx
Comment (evansiroky) [inline packages/trip-details/src/index.tsx:56]: I went ahead and tried using the `@formatjs/cli` cli tool and have verified that this approach is not able to properly generate a list of messages because that tool does not appear to do any runtime analysis of what is imported from the `../i18n/en-US.yml` file. It only really seemed to work if message strings were explicitly spelled out as full strings without any prefix or dynamic help. 

Therefore, I'm not convinced this is the exemplar approach that we should be using in every otp-ui file moving forward. The reason I used the `react-intl#defineMessages` method in otp-rr was because I thought if I didn't obtain that variable from the props in a functional component like in [this example](https://formatjs.io/docs/tooling/cli) that it wouldn't work. However, the cli tool does appear to pick up a signature of `intl.formatMessage({ id: "someId" })` anywhere, so I think the `defineMessages` thing is only useful when the same message ID would be used in multiple `intl.formatMessage` calls.

So, after this analysis, my recommendation is as follows:

- go ahead and still import the default messages, and I guess we have to flatten them because of jest (although it'd be ideal if we didn't have to)
- Remove the `expandAndDefineMessages` function and import of `react-intl#defineMessages`.
- In each FormattedMessage component 
  - spell out the entire flattened message ID in the `id` prop of a `FormattedMessage` 
  - spell out the entire flattened message ID in a lookup of the default message from the imported flattened default messages
  - if we are to keep the descriptions, also add a description prop

**Review by evansiroky — CHANGES_REQUESTED** [review]
See latest comment.

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: Thanks for clarifying the recommendations, I made the changes in https://github.com/opentripplanner/otp-ui/pull/281/commits/5f530377e60008a64f3408968fab91ec4f02e380.

Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: `@formatjs/cli` is interesting to add, I'll try that in a separate PR.

Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: https://github.com/ibi-group/otp-ui/pull/9 has a basic message id comparison between code and ymls using `@formatjs/cli`.

**Comment by binh-dam-ibigroup** [comment]
> There is potential to remove the `currency` prop that is introduced in this PR if #282 is adopted, and use that PR's currency calculation logic instead.

#282 is now adopted, so commit https://github.com/opentripplanner/otp-ui/pull/281/commits/506a8bd21c74ba893af7e8ce5aa1e9389f1c7113 removes the currency prop.

File: packages/trip-details/src/index.tsx
Comment (miles-grant-ibigroup) [inline packages/trip-details/src/index.tsx:2]: I would personally push for the approach used in #262 of putting the message in package.json so it's harder to forget about, but am sway-able

File: packages/trip-details/src/index.tsx
Comment (miles-grant-ibigroup) [inline packages/trip-details/src/index.tsx:27]: Do we want `<b>` instead of `<strong>`? Looking at the components, it seems like the bold text is semantically bold.

File: packages/trip-details/src/index.tsx
Comment (miles-grant-ibigroup) [inline packages/trip-details/src/index.tsx:81]: This may be unrelated to your work, but might be cleaner as a `filter()` and `map()`?

File: packages/trip-details/src/index.tsx
Comment (miles-grant-ibigroup) [inline packages/trip-details/src/index.tsx:133]: We seem to be using a few of the same props on a few of these `FormattedNumber`s. Do you think it's worth putting it into a helper method/wrapper component?

File: yarn.lock
Comment (miles-grant-ibigroup) [inline yarn.lock:2798]: When merging #285 into this, please make sure `moment-timezone` is out of the lock file!

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
This looks really great! Especially the Typescript architecture is looking really cool and looking forward to seeing everything else typed as well. I just have a few small questions, and then some stuff about merging #285

File: packages/trip-details/src/index.tsx
Comment (miles-grant-ibigroup) [reply packages/trip-details/src/index.tsx:81]: Or even a reducer

File: packages/trip-details/src/TripDetails.story.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/TripDetails.story.tsx:124]: Could you clarify your comment, please?

Comment (miles-grant-ibigroup) [reply packages/trip-details/src/TripDetails.story.tsx:124]: We're explicitly importing French or English messages based on the locale arg, and I was asking if this is story-specific or not, and if it is _not_ story specific if we could make it more dynamic. But I believe it's story-specific so shouldn't be an issue!

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: Good idea, refactored in https://github.com/opentripplanner/otp-ui/pull/281/commits/d39e81032da5df5266a7dfd0e664940c9cdbdd5f.

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: Refactored in https://github.com/opentripplanner/otp-ui/pull/281/commits/d39e81032da5df5266a7dfd0e664940c9cdbdd5f.

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: Right, changed to render `<strong>` in https://github.com/opentripplanner/otp-ui/pull/281/commits/d39e81032da5df5266a7dfd0e664940c9cdbdd5f.

File: packages/trip-details/src/TripDetails.story.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/TripDetails.story.tsx:124]: The locale and messages are story-specific (i.e. the implementer will have to provide the messages), and which messages gets passed to `<IntlProvider>` depends on the args set by the user in the Storybook Controls tab for that story. For simplicity, the contents of the YML files is statically loaded.

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: `<b>` tags replaced with `<strong>` in https://github.com/opentripplanner/otp-ui/pull/281/commits/d39e81032da5df5266a7dfd0e664940c9cdbdd5f.

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx:2]: I added the message to the `tsc` script of the trip-details package, in addition to the TypeScript exception comment which is required or a compilation error occurs. (see https://github.com/opentripplanner/otp-ui/pull/281/commits/d39e81032da5df5266a7dfd0e664940c9cdbdd5f)

File: yarn.lock
Comment (binh-dam-ibigroup) [reply yarn.lock]: Yep, moment-timezone is out after updating the core-utils dependency (see https://github.com/opentripplanner/otp-ui/pull/281/commits/7735f8b5d76bee70fe23841ee6a9d01a34da5238)

File: packages/trip-details/package.json
Comment (binh-dam-ibigroup) [reply packages/trip-details/package.json]: I brought updates from #285, it was simpler to resolve conflicts from yarn.lock.

**Comment by binh-dam-ibigroup** [comment]
Ready for review again @miles-grant-ibigroup, but I would wait for #285 to be merged before merging this one.

File: packages/trip-details/src/index.tsx
Comment (miles-grant-ibigroup) [reply packages/trip-details/src/index.tsx:2]: Really like this approach thanks!

**Review by miles-grant-ibigroup — APPROVED** [review]
Looks really great! Thanks so much for all your work on moving this forward.

File: packages/trip-details/src/types.ts
Comment (fpurcell) [inline packages/trip-details/src/types.ts:5]: This should be fixed before merging, as the note implies.  Will either @opentripplanner/types/src  -or- #/types/src work?

**Review by fpurcell — CHANGES_REQUESTED** [review]
I took a brief look.  I feel like this will work for the TORA project.  I'm good with merging now, and going from there ... would like two minor issues addressed first tho, Binh.

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx:121]: @fpurcell Fares returned from OTP are in cents. Previously, TripDetails was using a utility in the OTP-UI package to convert cents to dollars, which we replaced to accommodate the currency returned in the itinerary.

Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx:121]: Placing formatting tags inside the resource strings makes the most sense, especially in the expandable details for the calories where the order/number of words to be formatted can vary.

File: packages/trip-details/src/types.ts
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/types.ts]: Thanks for pointing out this one. `@opentripplanner/types` seems to work also so I used that in https://github.com/opentripplanner/otp-ui/pull/281/commits/1fa98e123757bdfce2c79781f885d8261283c21a.

Comment (binh-dam-ibigroup) [reply packages/trip-details/src/types.ts]: After changing the import to use the "cleaner" syntax, it looks like the `types` package needs to be published first for the CI to pass, so I opened #291.

**Comment by binh-dam-ibigroup** [comment]
Blocking this PR pending #291 or alternate advice for https://github.com/opentripplanner/otp-ui/pull/281#discussion_r713417109.

File: packages/trip-details/src/types.ts
Comment (fpurcell) [inline packages/trip-details/src/types.ts:14]: See notes above about getting rid of moment as a dependency.

Comment (binh-dam-ibigroup) [reply packages/trip-details/src/types.ts:13]: Good suggestion, thanks, removed in https://github.com/opentripplanner/otp-ui/pull/281/commits/a9adf69db199e0306ae19215a8bd88625b28634f.

File: packages/trip-details/src/index.tsx
Comment (binh-dam-ibigroup) [reply packages/trip-details/src/index.tsx]: Removed in https://github.com/opentripplanner/otp-ui/pull/281/commits/a9adf69db199e0306ae19215a8bd88625b28634f.

**Review by grant-humphries — APPROVED** [review]
All looks good here to me and I think the issues that Frank raised have been addressed, are we ready to merge?

**Comment by binh-dam-ibigroup** [comment]
> All looks good here to me and I think the issues that Frank raised have been addressed, are we ready to merge?

@grant-humphries Almost! There is a dependency in #291 for the build, and @fpurcell has the blocking requested change for this PR.

**Comment by fpurcell** [comment]
Apologies @binh-dam-ibigroup, I didn't realize I was a blocker till this morning.  (feel free to ping on slack if I don't respond via github -- that should get my attention)

**Pull request merged**
